### ISSUE TITLE
[libtorch] Is there any global lock in process?

### ISSUE THREAD

**Issue by wizardk** [pytorch/pytorch#26948]
## 🐛 Bug

CPU usage is not increasing by increasing the number of threads when inferring with CPU only libtorch in one process and multiple threads.

## To Reproduce

```
const int THREAD_CNT = 10

torch::jit::script::Module load()
{
    torch::NoGradGuard no_guard;
    torch::jit::script::Module module = torch::jit::load("xxx.jit");
    module.to(torch::kCPU);
    module.eval();
    return module;
}

torch::TensorOptions options = torch::TensorOptions().dtype(torch::kFloat32).requires_grad(false).device(torch::kCPU);

void * working_thread(void * param)
{
    torch::init_num_threads();
    torch::jit::script::Module * module = (torch::jit::script::Module *)param;
    torch::NoGradGuard no_guard;
    while (1) {
        torch::Tensor x = torch::randn(1, 10, options);
        torch::Tensor y = module->forward({x}).toTensor();
    }
    return 0;
}

void main()
{
    // load module
    torch::jit::script::Module module = load();

    // muti-threading
    pthread_t thread_handle[THREAD_CNT];
    for (int thread_idx = 0; thread_idx < THREAD_CNT; ++thread_idx) {
        pthread_create(thread_handle + thread_idx, 0, working_thread, &module);
    }

    // finished
    for (int thread_idx = 0; thread_idx < THREAD_CNT; ++thread_idx) {
        pthread_join(thread_handle[thread_idx], 0);
    }
}
```

## Expected behavior

One process, more threads, higher CPU Usage.

## Environment

 - LibTorch Version : 1.2
 - OS : Linux
 - How you installed LibTorch : download from pytorch.org
 - CPU : 96 cores

## Additional context

Here is my experiment result:
I can get 33% CPU usage by 1 process and 1 thread per process.
I can get 66% CPU usage by 1 process and 2 threads per process.
I can get 46% CPU usage by 1 process and 3 threads per process.
I can get 74% CPU usage by 2 processes and 1 thread per process.
I can get 92% CPU usage by 2 processes and 2 threads per process.

I guess there should be a global lock in process range to limit the performance of multi-threading inference.

cc @suo

**Comment by wizardk** [comment]
@pbelevich @izdeby Moreover, I found that the first inference is significantly slower than the following inference.

**Comment by soumith** [comment]
@wizardk the first inference will be slower due to state initialization across the library

**Comment by wizardk** [comment]
@soumith Thanks. So this is a feature, not a bug, right? But it still is a little confusing. Does the first state initialization need an input case? I can understand if so. Contrarily, what about exposing an API for caller to do this initialization?

**Comment by suo** [comment]
The first inference is slower as well due to the initial optimization that the JIT is doing on your code. This is similar to "warm up" in other JIT compilers. Typically, production services will warm up a model using representative inputs before marking it as available.

**Comment by wizardk** [comment]
@suo Thanks.

**Comment by suo** [comment]
However, there is no global lock in libtorch that prevents multithreaded inference from scaling reasonably. If you have more detailed profiles I would be very interested in checking them out.

**Comment by wizardk** [comment]
1. Make a jit model
```
class MyModule(torch.nn.Module):
    def __init__(self):
        super(MyModule, self).__init__()
        self.lyr = [torch.nn.Linear(1000, 1000) for _ in range(100)]
        self.lyr = torch.nn.Sequential(*self.lyr)

    def forward(self, x):
        return self.lyr(x)

my_module = MyModule()
my_module = torch.jit.script(my_module)
my_module.save('traced_model.pt')
```

2. Run in cpp
```
#include <stdio.h>
#include <stdlib.h>
#include <torch/script.h>
#include <torch/torch.h>

torch::TensorOptions options = torch::TensorOptions().dtype(torch::kFloat32).requires_grad(false).device(torch::kCPU);
int thread_cnt, loop_cnt;

torch::jit::script::Module load()
{
    torch::NoGradGuard no_guard;
    torch::jit::script::Module module = torch::jit::load("traced_model.pt");
    module.to(torch::kCPU);
    module.eval();
    return module;
}

void * working_thread(void * param)
{
    torch::init_num_threads();
    torch::jit::script::Module * module = (torch::jit::script::Module *)param;
    for (int idx = 0; idx < loop_cnt; ++idx) {
        torch::NoGradGuard no_guard;
        torch::Tensor x = torch::randn({1, 1000}, options);
        torch::Tensor y = module->forward({x}).toTensor();
    }
    return 0;
}

int main(int argc, char ** argv)
{
    if (argc != 3) {
        return 0;
    }
    thread_cnt = atoi(argv[1]);
    loop_cnt = atoi(argv[2]);

    torch::jit::script::Module module = load();

    pthread_t * thread_handles = (pthread_t *)malloc(sizeof(pthread_t) * thread_cnt);
    for (int idx = 0; idx < thread_cnt; ++idx) {
        pthread_create(thread_handles + idx, 0, working_thread, &module);
    }
    for (int idx = 0; idx < thread_cnt; ++idx) {
        pthread_join(thread_handles[idx], 0);
    }
    free(thread_handles);

    return 0;
}
```

3. Result
./test 1 10000 => CPU usage 2400
./test 2 10000 => CPU usage 4800
./test 3 10000 => CPU usage 2700
./test 4 10000 => CPU usage 3000
./test 5 10000 => CPU usage 3100

@suo There are 48 CPU cores and torch::get_num_threads() returns 24.
Is this test result normal? It seems there was a sudden drop in the performance of mkldnn when the total number of threads used in parallel region exceeds the number of CPU cores.

**Comment by suo** [comment]
I suspect this has to do with OpenMP contention; can you try running your benchmark with `OMP_NUM_THREADS=1` set?

**Comment by wizardk** [comment]
export OMP_NUM_THREADS=1

1 process:
/test 1 10000 => CPU usage : 100
/test 2 10000 => CPU usage : 200
/test 3 10000 => CPU usage : 300
/test 4 10000 => CPU usage : 400
/test 5 10000 => CPU usage : 500

2 processes:
./test 1 10000 => CPU usage : 100+100
./test 2 10000 => CPU usage : 200+200
./test 3 10000 => CPU usage : 300+300
./test 4 10000 => CPU usage : 400+400
./test 5 10000 => CPU usage : 500+500

@suo

**Comment by wizard-baller** [comment]
I  cared for CPU usage in the last time. Now, I did another speed test on libtorch 1.3.0.
The test code is almost same with the previous. I only add some code to compute elapsed time by std::chrono::system_clock. And there are 48 CPU cores.

The test result are:

set_num_interop_threads(1),  set_num_threads(1), 1 process, 1 working thread, 100 times inference in each thread:
average inference time is 536 ms, CPU usage is 100.

set_num_interop_threads(1),  set_num_threads(1), 1 process, 10 working threads, 100 times inference in each thread:
average inference time is 749 ms, CPU usage is 1000.

set_num_interop_threads(1),  set_num_threads(1), 1 process, 20 working threads, 100 times inference in each thread:
average inference time is 1363 ms, CPU usage is 2000.

set_num_interop_threads(1),  set_num_threads(1), 1 process, 40 working threads, 100 times inference in each thread:
average inference time is 7223 ms, CPU usage is 4000.

You can find out the CPU usage looks normal but inference speed is abnormal.
@suo Any progress on the CPU usage issue? There is another speed issue has nothing to do with OpenMP I guess.

**Comment by wizardk** [comment]
I cared for CPU usage in the last time. Now, I did another speed test on libtorch 1.3.0.
The test code is almost same with the previous. I only add some code to compute elapsed time by std::chrono::system_clock. And there are 48 CPU cores.

The test result are:

1.
set_num_interop_threads(1), set_num_threads(1), 1 process, 1 working thread, 100 times inference in each thread:
average inference time is 536 ms, CPU usage is 100.

2.
set_num_interop_threads(1), set_num_threads(1), 1 process, 10 working threads, 100 times inference in each thread:
average inference time is 749 ms, CPU usage is 1000.

3.
set_num_interop_threads(1), set_num_threads(1), 1 process, 20 working threads, 100 times inference in each thread:
average inference time is 1363 ms, CPU usage is 2000.

4.
set_num_interop_threads(1), set_num_threads(1), 1 process, 40 working threads, 100 times inference in each thread:
average inference time is 7223 ms, CPU usage is 4000.

You can find out the CPU usage looks normal but inference speed is abnormal.
@suo Any progress on the CPU usage issue? There is another speed issue has nothing to do with OpenMP I guess.

**Comment by wizardk** [comment]
I cared for CPU usage in the last time. Now, I did another speed test on libtorch 1.3.0.
The test code is almost same with the previous. I only add a warm-up step before creating threads and add some code to compute elapsed time by std::chrono::system_clock. And there are 48 CPU cores.

The test code is:
```
#include <torch/script.h>
#include <torch/torch.h>
#include <iostream>
#include <chrono>
#include <vector>
#include <numeric>

torch::TensorOptions options = torch::TensorOptions().dtype(torch::kFloat32).requires_grad(false).device(torch::kCPU);
torch::jit::script::Module module;
int omp_cnt, omp_inter_cnt, thread_cnt, loop_cnt;

torch::jit::script::Module load()
{
    torch::NoGradGuard no_guard;

    torch::jit::script::Module module = torch::jit::load("model.jit");
    module.to(torch::kCPU);
    module.eval();
    return module;
}

void warmup()
{
    torch::NoGradGuard no_guard;
    std::chrono::system_clock::time_point start = std::chrono::system_clock::now();
    torch::Tensor x = torch::randn({1, 80, 100}, options);
    torch::Tensor y = module.forward({x}).toTensor();
    std::chrono::milliseconds elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now() - start);
    std::cout << "warmup " << elapsed.count() << std::endl;
}

void * working_thread(void * param)
{
    torch::init_num_threads();
    if (omp_cnt > 0) {
        torch::set_num_threads(omp_cnt);
    }
    std::cout << "thread" << std::endl;
    std::cout << "torch::get_num_threads = " << torch::get_num_threads() << std::endl;
    std::cout << "torch::get_num_interop_threads = " << torch::get_num_interop_threads() << std::endl;
    
    int * ms = (int *)param;
    for (int idx = 0; idx < loop_cnt; ++idx) {
        torch::NoGradGuard no_guard;
        torch::Tensor x = torch::randn({1, 80, 100}, options);
        std::chrono::system_clock::time_point start = std::chrono::system_clock::now();
        torch::Tensor y = module.forward({x}).toTensor();
        std::chrono::milliseconds elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now() - start);
        *ms += elapsed.count();
    }
    *ms /= loop_cnt;
    std::cout << "thread quit" << std::endl;
    return 0;
}

int main(int argc, char ** argv)
{
    if (argc != 5) {
        return 0;
    }
    omp_cnt = atoi(argv[1]);
    omp_inter_cnt = atoi(argv[2]);
    thread_cnt = atoi(argv[3]);
    loop_cnt = atoi(argv[4]);
    std::cout << "main" << std::endl;
    std::cout << "torch::get_num_threads = " << torch::get_num_threads() << std::endl;
    std::cout << "torch::get_num_interop_threads = " << torch::get_num_interop_threads() << std::endl;
    if (omp_inter_cnt > 0) {
        torch::set_num_interop_threads(omp_cnt);
    }

    module = load();
    warmup();

    pthread_t * thread_handles = new pthread_t [thread_cnt];
    std::vector<int> ms(thread_cnt, 0);
    for (int idx = 0; idx < thread_cnt; ++idx) {
        pthread_create(thread_handles + idx, 0, working_thread, &ms[idx]);
    }
    for (int idx = 0; idx < thread_cnt; ++idx) {
        pthread_join(thread_handles[idx], 0);
    }
    delete [] thread_handles;

    for (int idx = 0; idx < ms.size(); ++idx) {
        std::cout << "ms[" << idx << "]" << ms[idx] << std::endl;
    }
    std::cout << std::accumulate(ms.begin(), ms.end(), 0) / ms.size() << std::endl;
    
    return 0;
}
```

The test result are:

set_num_interop_threads(1), set_num_threads(1), 1 process, 1 working thread, 100 times inference in each thread:
average inference time is 463 ms, CPU usage is 100.

set_num_interop_threads(1), set_num_threads(1), 1 process, 10 working threads, 100 times inference in each thread:
average inference time is 677 ms, CPU usage is 1000.

set_num_interop_threads(1), set_num_threads(1), 1 process, 20 working threads, 100 times inference in each thread:
average inference time is 1046 ms, CPU usage is 2000.

set_num_interop_threads(1), set_num_threads(1), 1 process, 40 working threads, 100 times inference in each thread:
average inference time is 6083 ms, CPU usage is 4000.

You can find out the CPU usage looks normal but inference speed is abnormal.
@suo Any progress on the CPU usage issue? There is another speed issue has nothing to do with OpenMP I guess.

**Comment by wizardk** [comment]
@suo I've tried to discard jit and turn to infer by layer(torch::nn::xxx) and function, but the result is still bad. What is the linear algebra accelerating backend of precompiled version?

**Comment by wizardk** [comment]
Moreover, I have tested onnxruntime, it has similar behavior of rapid deterioration of performance.

**Comment by suo** [comment]
It does indeed sound like an issue in the underlying math libraries. I'm not sure exactly which ones would be invoked here, perhaps running things through perf would yield interesting results?

**Comment by suo** [comment]
cc @jamesr66a: do you know who would know about contention in MKL or whatever we're using here

**Comment by jamesr66a** [comment]
@wizardk could you provide the `model.jit` file used in your most recent example? It's hard to diagnose what's going on without it

I did some experiments with the original traced model and the new C++ code and the performance characteristics do not look out of the ordinary on my machine. The majority of cycles are spent in MKL BLAS sgemm or gemv routines. My best hypothesis for the performance degradation at this point is as more threads are added doing parallel inference on the same parameters, more memory traffic is generated and that causes saturation DRAM/cache bandwidth saturation much earlier than we reach full CPU subscription. This is especially prevalent for the small batch size case, as in the original examples. Unfortunately, this is an inherent limitation of modern computer architecture.

We do indeed have production teams that run floating-point models at only ~30% CPU subscription because of this very issue. One way to improve the situation is to use `quantization`, see https://pytorch.org/docs/stable/quantization.html. This will reduce the size of the parameters and lower the demand on the memory subsystem

**Comment by wizardk** [comment]
@jamesr66a Thanks for your helpful reply. I did a memory traffic experiment and it turned out to be the same as you said. Bigger continuous memory, more parallel tasks, slower execution speed. I will try to compress model in CPU mode or move the parameters to GPU.

@suo Thanks for your help as well.

**Comment by oelgendy** [comment]
> The first inference is slower as well due to the initial optimization that the JIT is doing on your code. This is similar to "warm up" in other JIT compilers. Typically, production services will warm up a model using representative inputs before marking it as available.

In my code, I only do the first inference, then obtain network output from the exe file. Is there a way to make the first inference faster if I run the exe file multiple times?

**Comment by suo** [comment]
Unfortunately that's not possible today, as we perform various optimizations based on the input during the first inference. We are looking to improve first inference speed by doing the specialization/optimization off-thread, but that isn't yet available.
